Explore: Change Logs.tsx to a functional component

gelicia opened the pull request:
**What is this feature?**

This changes Logs.tsx to be a functional component. There should be no visible difference to the end user.

**Why do we need this feature?**

Converting this component to be functional will help further improvements. See https://github.com/grafana/grafana/pull/86431#issuecomment-2082612302 for more details.

**Which issue(s) does this PR fix?**:

Fixes #87047

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

gelicia committed:
First pass

gelicia committed:
WIP

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into kristina/logs-functional

gelicia committed:
why is my cpu angy

ifrost committed:
Avoid triggering onHiddenSeriesChanged too often

onHiddenSeriesChanged should be called only when dataWithConfig changes, not when the callback changes itself.

This is actually causing an infinite loop because onHiddenSeriesChanged may not be memoized in the parent and passed as a  new callback function on each render.

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into kristina/logs-functional

gelicia committed:
fix tests

gelicia committed:
Remove store mock and clear store between tests instead

gelicia committed:
Fix filtering

matyax commented on public/app/features/explore/Logs/Logs.tsx:
```diff
@@ -156,81 +140,88 @@ const getDefaultVisualisationType = (): LogsVisualisationType => {
+  let cancelFlippingTimer: number | undefined = undefined;
   // @ts-ignore
-  private toggleLegendRef: React.MutableRefObject<(name: string, mode: SeriesVisibilityChangeMode) => void> =
+  const toggleLegendRef: React.MutableRefObject<(name: string, mode: SeriesVisibilityChangeMode) => void> =
     React.createRef();
+  const topLogsRef = React.createRef<HTMLDivElement>();
```
I think these could be `useRef()`.

ifrost replied:
Fixed in 06dcdc8ca9d7344131a2f3e0ed081f463bde4a08

matyax commented on public/app/features/explore/Logs/Logs.tsx:
```diff
@@ -248,84 +239,103 @@ class UnthemedLogs extends PureComponent<Props, State> {
+        );
+      }
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Disabling exhaustive deps should be an exceptional measure of last resort, so in general in this PR I would ask you to not use it.

For this particular case, the clean up function will be created and returned with values that might not be valid or present when it's executed, specially those pointing to timers.

ifrost replied:
Good catch! We may need to use `useRef` to store latest `panelState.logs` and call it only on unmount. In long-run this code looks like should not be here at all and Explore should clean up panel state for each panel automatically 🤔

ifrost replied:
Fixed in effed375e1bfadcaec7ed926fed926d3445f8e05 + a75b5c80be0994f09e3ea7987d867acc65b52633.

matyax commented on public/app/features/explore/Logs/Logs.tsx:
```diff
@@ -248,84 +239,103 @@ class UnthemedLogs extends PureComponent<Props, State> {
     }
-    if (this.props.panelState?.logs?.visualisationType !== prevProps.panelState?.logs?.visualisationType) {
-      const visualisationType = this.props.panelState?.logs?.visualisationType ?? getDefaultVisualisationType();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Same here. Forcing the dependencies to not update should not be the solution, because the point of the dependency array and the linting rule is to use valid referenced values and prevent bugs.

matyax commented on public/app/features/explore/Logs/Logs.tsx:
```diff
@@ -640,353 +622,299 @@ class UnthemedLogs extends PureComponent<Props, State> {
       }
     }
-    this.topLogsRef.current?.scrollIntoView();
+    topLogsRef.current?.scrollIntoView();
```
Aprox from line 314 to 634, these used to be class methods that would not be recreated on every render. Could you please wrap them around `useCallback()`. The reason is to maintain the previous behavior and to not cause performance regressions and/or unnecessary re-renders.

ifrost replied:
Made memoized utils static in 5d2f0ebd841c3343665712e73099f4a6e38f3f37 and added wrappers in fed463c46154ba0ac7442c745b21066e9b2b4d95

matyax reviewed (commented):
Thanks a lot for taking care of this refactor!

In general this is a very big change that doesn't have enough previous testing coverage to guide it. In general what worries me the most is the problem with dependency arrays. Disabling it will cause the effects and clean up functions to use potentially stale references with unexpected results, so it gives me very little confidence to release it as it is. I would greatly appreciate if you can work around them.

If I can help with this, I'm happy to do it, as this is a key component for logs in Explore.

gelicia committed:
First pass adding callbacks and dependencies

harisrozajac committed:
Merge branch 'main' into kristina/logs-functional

harisrozajac committed:
Add useRef hook for toggleLegend and topLogsRef

ifrost committed:
Remove unwanted reactivity

register(...) changes outlineItems. A function that calls register(...) behaves like a setter for the state - it cannot react to the state itself.

ifrost committed:
Separate unmount clean-ups

ifrost committed:
Ensure unmount is not reactive to props

ifrost committed:
Make memoized functions stats

ifrost committed:
Wrap functions passed down to components with useCallback

ifrost committed:
Fix reporting interaction when context is closed

harisrozajac reviewed (approved):
This looks great to me. I've tested it with content outline and it works as expected. Also tested with loki, elastic, and cloudwatch and didn't find any issues. Great work at this @gelicia and @ifrost 🥇

ifrost commented:
@matyax / @grafana/observability-logs -> it's ready for another round of review!

matyax commented on public/app/features/explore/Logs/Logs.tsx:
```diff
@@ -156,81 +141,90 @@ const getDefaultVisualisationType = (): LogsVisualisationType => {
+
+  const logsVolumeEventBus = eventBus.newScopedBus('logsvolume', { onlyLocal: false });
+  const { register, unregisterAllChildren } = useContentOutlineContext() ?? {};
+  let flipOrderTimer: number | undefined = undefined;
+  let cancelFlippingTimer: number | undefined = undefined;
```
```suggestion
  let flipOrderTimer: number | undefined = undefined;
  let cancelFlippingTimer: number | undefined = undefined;
```

Shouldn't these be refs? I have no way of telling if `useUnmount` would receive the expected values when it executes. Worst case scenario, using refs would have no unwanted side effects.

ifrost replied:
useUnmount uses refs under the hood - https://github.com/streamich/react-use/blob/master/src/useUnmount.ts

matyax replied:
Right, but on every render() call `flipOrderTimer` and `cancelFlippingTimer` will be created with undefined value, potentially overriding (or forgetting) the previous value. Before we used class attributes for that, which hold the same value as long as the component instance lives.

harisrozajac replied:
Addressed in https://github.com/grafana/grafana/pull/87808/commits/55b82654ab0c6dbffef5a3164c486829a288a301

matyax commented on public/app/features/explore/Logs/Logs.tsx:
```diff
@@ -156,81 +141,90 @@ const getDefaultVisualisationType = (): LogsVisualisationType => {
+  const [visualisationType, setVisualisationType] = useState<LogsVisualisationType | undefined>(
+    panelState?.logs?.visualisationType ?? getDefaultVisualisationType()
+  );
+  const [logsContainer, setLogsContainer] = useState<HTMLDivElement | undefined>(undefined);
```
This doesn't need to be a state variable, and also by the way it's implemented we can just use a ref.

```
const onLogsContainerRef = useCallback((node: HTMLDivElement) => {
    setLogsContainer(node);
  }, []);
```

harisrozajac replied:
Addressed in https://github.com/grafana/grafana/pull/87808/commits/55b82654ab0c6dbffef5a3164c486829a288a301

matyax commented on public/app/features/explore/Logs/Logs.tsx:
```diff
@@ -248,319 +242,334 @@ class UnthemedLogs extends PureComponent<Props, State> {
+    setDisplayedFields([]);
+  }, []);
+
+  const [onCloseCallback, setOnCloseCallback] = useState<() => void>(() => {});
```
This can also just be a ref, as changing the value has no rendering consequence.

harisrozajac replied:
Addressed in https://github.com/grafana/grafana/pull/87808/commits/55b82654ab0c6dbffef5a3164c486829a288a301

matyax reviewed (commented):
Amazing work. Left some comments, but looks great.

harisrozajac committed:
Merge branch 'main' into kristina/logs-functional

harisrozajac committed:
Adjust several variables to use refs

matyax commented:
This should be ready to merge. If I may suggest, this friday is the 11.1 cutoff, so I would wait after that date to merge, just to be on the safe side and prevent having to do backports for any unexpected bug. What do you think?

ifrost committed:
Merge branch 'refs/heads/main' into kristina/logs-functional

# Conflicts:
#	public/app/features/explore/ContentOutline/ContentOutlineContext.tsx
#	public/app/features/explore/Logs/Logs.tsx

ifrost committed:
Post-merge fixes

ifrost committed:
Merge branch 'refs/heads/main' into kristina/logs-functional

ifrost commented:
Back to the draft. Scrolling to pinned line doesn't seem to work properly.

ifrost committed:
Merge branch 'refs/heads/main' into kristina/logs-functional

ifrost committed:
Ensure scrollIntoView is called with a ref to the container

ifrost committed:
Update docs

ifrost commented:
Please check the latest change to `onLogsContainerRef` to make scrolling to pinned logs working.

diegommm commented:
Hi! Greetings from the Release Guild. Just letting you know I'm updating your PR's milestone to [11.2.x](https://github.com/grafana/grafana/milestone/529). If you need to see your changes in a previous version, please consider adding a backport label.

Thank you!

ifrost committed:
Ensure scrollIntoView is updated only if the scroll element changes

ifrost committed:
Fix the default pinned log tooltip

ifrost committed:
Small clean up

harisrozajac committed:
Remove console.log()

harisrozajac reviewed (approved):
Looks great 🥇 I checked that scrolling works when:
- Pasting a URL created by "Copy shortlink"
- Clicking on a pinned log

matyax commented:
Tomorrow I'll be taking another look at this one. Thanks for your patience!

matyax reviewed (approved):
Great work! Thanks for making this change and iterating on it.

ifrost commented on public/app/features/explore/Logs/Logs.tsx:
```diff
@@ -597,455 +635,347 @@ class UnthemedLogs extends PureComponent<Props, State> {
+                  permalinkedRowId={panelState?.logs?.id}
+                  scrollIntoView={scrollIntoView}
+                  isFilterLabelActive={props.isFilterLabelActive}
+                  containerRendered={!!logsContainerRef}
```
This may be also worth revising. I think it will always be truthy (the ref is always created, only .current changes). But with latest changes to setting scrollIntoView only once the ref is received it may not be needed at all.
